### DISCUSSION TITLE
Boot Process, Flat Device Tree, mkbootimg Discussion

### DISCUSSION THREAD

**Discussion by mj-sakellaropoulos** [hansemro/kc1-linux#37] [General]
> Original line of question : Where is device tree stored?
>
> mj-sakellaropoulos 4 hours ago
> 
> Devicetree location:
> Method 1 : Appended after zImage in kernel binary (eg: cat zImage omap4-kc1.dtb)
> Uboot will look for the magic signature of device tree binary and pass to pass the dtb to the kernel ( i think )
> Method 2 : Android mkbootimg --dt / inside android bootimg
> Android boot.img contains a header with all the information of subcomponents offsets, uboot parses this header and passes the devicetree to kernel ( i think )
> 
>  
> hansemro 17 minutes ago Author Owner
> 
> I don't know if it is still true, but osm0sis' fork of mkbootimg (which pmbootstrap uses) doesn't support device trees properly.
> 
> Does method 1 work for you?
> 
>  
> hansemro 15 minutes ago Author Owner
> 
> I ask where because I was having doubts how pmbootstrap packages the device tree. Is it installed with device-amazon-otter*.apk? Does it come from linux-amazon-otter*.apk? Or does it come with the rootfs image?

I am using mkbootimg from here: https://github.com/CyanogenMod/android_system_core.git

With this version, if I run `abootimg -i boot.img` against an image that it generated, it's considered invalid. However it's working with Hashcode's u-boot and fastboot commands

**Comment by mj-sakellaropoulos** [comment]
> Does method 1 work for you?
Yes
> I ask where because I was having doubts how pmbootstrap packages the device tree. Is it installed with device-amazon-otter*.apk? Does it come from linux-amazon-otter*.apk? Or does it come with the rootfs image?

I'm not familiar with the version that pmOS uses

It seems google has move to a python implementation

As far as I know, bootloader is responsible of passing the "Flattened device tree" to the kernel

**Reply by hansemro** [reply]
> bootloader is responsible of passing the "Flattened device tree" to the kernel

Yes, that is correct. However, the version of U-Boot we are using does not seem to comply with some Android boot image standards, so it fails to find and load the device tree for certain boot images. Because I did not want to deal with Android boot image nonsense in early development, I started out by loading the kernel and device tree manually with U-Boot commands.

**Reply by hansemro** [reply]
I have a small guide on how I load and boot the kernel with device tree with U-boot from eMMC in the README.
https://github.com/hansemro/kc1-linux#manual-method

**Reply by hansemro** [reply]
What I left out from the guide was how to load the kernel over serial since I figured it made more sense to load the device tree and kernel over USB in recovery mode.

**Reply by hansemro** [reply]
> I'm not familiar with the version that pmOS uses

pmOS uses osm0sis fork: http://pkgs.postmarketos.org/packages?name=mkbootimg*&branch=master&origin=

You can try it out here: https://github.com/osm0sis/mkbootimg

**Comment by hansemro** [comment]
Small tutorial on loading device tree over serial
------------------

Use kermit tool (such as `ukermit`) to send fdt (https://github.com/nmenon/omap-u-boot-utils)

Enter kermit download mode
```
U-Boot# loadb $(fdtaddr)
```
Send dtb file with ukermit
```
$ ./ukermit -p /dev/ttyUSB0 -f ./omap4-kc1.dtb
Downloading file: 100.000% completed (51162/51162 bytes)
File Download completed
```
Read test:
```
U-Boot# fdt addr $(fdtaddr)
U-Boot# fdt list /
/ {
        #address-cells = <0x00000001>;
        #size-cells = <0x00000001>;
        compatible = "amazon,omap4-kc1", "ti,omap4430", "ti,omap4";
        interrupt-parent = <0x00000001>;
        model = "Amazon Kindle Fire (first generation)";
        chosen {
        };
        aliases {
        };
        memory {
        };
        cpus {
        };
        interrupt-controller@48241000 {
        };
        l2-cache-controller@48242000 {
        };
        local-timer@48240600 {
        };
        interrupt-controller@48281000 {
        };
        soc {
        };
        ocp {
        };
        thermal-zones {
        };
};
U-Boot# fdt list /cpus
cpus {
        #address-cells = <0x00000001>;
        #size-cells = <0x00000000>;
        cpu@0 {
        };
        cpu@1 {
        };
};
U-Boot# fdt list /cpus/cpu@0
cpu@0 {
        compatible = "arm,cortex-a9";
        device_type = "cpu";
        next-level-cache = <0x00000002>;
        reg = <0x00000000>;
        clocks = <0x00000003>;
        clock-names = "cpu";
        clock-latency = <0x000493e0>;
        operating-points = <0x000493e0 0x000fa3e8 0x000927c0 0x00124f80 0x000c3500 0x001408e8 0x000f6180 0x0014fb18>;
        cooling-min-level = <0x00000000>;
        cooling-max-level = <0x00000003>;
        #cooling-cells = <0x00000002>;
        linux,phandle = <0x00000092>;
        phandle = <0x00000092>;
};
```
